lib/saluki-components/src/transforms/aggregate/mod.rs

```diff
@@ -290,9 +286,21 @@ impl Transform for Aggregate {
 
                         for event in events {
                             if let Some(metric) = event.try_into_metric() {
-                                if self.forward_timestamped_metrics && metric.metadata().timestamp().is_some() {
-                                    event_buffer.push(Event::Metric(metric));
-                                } else if !state.insert(current_time, metric) {
+                                let metric = if self.forward_timestamped_metrics {
+                                    // If we're configured to forward timestamped metrics immediately, then we need to
+                                    // try to handle any timestamped values in this metric. If we get back `Some(...)`,
+                                    // it's either the original metric because no values had timestamps _or_ it's a
+                                    // modified version of the metric after all timestamped values were split out and
+                                    // directly forwarded.
+                                    match handle_forward_timestamped_metric(metric, &mut event_buffer) {
```

Comment:
This is making me wonder if it'd be better to split out timestamped metric samples earlier (i.e. at the source) and send them through a different pipeline that doesn't include the aggregate transform. That'd be a much more direct mapping of the agent's current no-aggregation "pipeline". We could even differentiate more strongly between "samples" (single data point, no timestamp, to be aggregated) and "metrics" (potentially multiple points, timestamped, aggregated) at the topology level and have different components that work on each (e.g. `aggregate` is `samples -> metrics`). Just a thought.

Reply:
The other bit, which the Datadog Agent gets around by deconstructing what we have with the `MetricValue` enum, is by having all the non-common data -- rate interval, etc -- on the metric sample type, and then just having the raw samples in a vector.

Our choice here of having `MetricValue` definitely paints us into a corner, to an extent, since it makes it harder to have homogenous value containers.